src/System.Net.Primitives/src/System/Net/IPAddress.cs

```diff
@@ -542,33 +544,31 @@ public long Address
             }
         }
 
-        internal bool Equals(object comparandObj, bool compareScopeId)
+        /// <summary>Compares two IP addresses.</summary>
+        public override bool Equals(object comparand)
         {
-            IPAddress comparand = comparandObj as IPAddress;
+            return comparand is IPAddress address && Equals(address);
+        }
 
-            if (comparand == null)
-            {
-                return false;
-            }
+        internal bool Equals(IPAddress comparand)
+        {
+            Debug.Assert(comparand != null);
 
             // Compare families before address representations
             if (AddressFamily != comparand.AddressFamily)
             {
                 return false;
             }
+
             if (IsIPv6)
             {
-                // For IPv6 addresses, we must compare the full 128-bit representation.
-                for (int i = 0; i < NumberOfLabels; i++)
-                {
-                    if (comparand._numbers[i] != _numbers[i])
-                    {
-                        return false;
-                    }
-                }
-
-                // The scope IDs must also match
-                return comparand.PrivateScopeId == PrivateScopeId || !compareScopeId;
+                // For IPv6 addresses, we must compare the full 128-bit representation and the scope IDs.
+                ReadOnlySpan<byte> thisNumbers = MemoryMarshal.AsBytes<ushort>(_numbers);
+                ReadOnlySpan<byte> comparandNumbers = MemoryMarshal.AsBytes<ushort>(comparand._numbers);
+                return
+                    MemoryMarshal.Read<ulong>(thisNumbers) == MemoryMarshal.Read<ulong>(comparandNumbers) &&
+                    MemoryMarshal.Read<ulong>(thisNumbers.Slice(sizeof(ulong))) == MemoryMarshal.Read<ulong>(comparandNumbers.Slice(sizeof(ulong))) &&
+                    PrivateScopeId == comparand.PrivateScopeId;
```

Comment:
Wondering if the intrinsics API can be used here 🙂 
```csharp
if (Sse41.IsSupported)
{ 
    fixed (ushort* num1ptr = _numbers)
    fixed (ushort* num2ptr = comparand._numbers)
    {
        return Sse41.TestAllOnes( // or Sse2.MoveMask(...) == 0xFFFF whatever is faster
            Sse2.CompareEqual(Sse2.LoadVector128(num1ptr), Sse2.LoadVector128(num2ptr)));
    }
}
```

Reply:
Potentially.  You're welcome to try after this merges.  My goal wasn't to make it as fast as possible, only as fast as it already was but now with an additional check.  I'm also not sure it's worth adding unsafe code for this.

Reply:
Anyway it seems your implementation is better anyway, SSE is a bit faster in cases when two IPs are equal but significantly slower in other cases.

Reply:
Thanks for checking :)